lib/ui/painting.dart

```diff
@@ -4263,6 +4272,111 @@ class FragmentProgram extends NativeFieldWrapperClass1 {
+/// set by calling [setSampler].
+///
+/// A [FragmentShader] can be re-used, and this is an efficient way to avoid
+/// allocating and re-initializing the uniform buffer and samplers. However,
+/// if two [FragmentShader] objects with different float uniforms or samplers
+/// are required to exist simultaneously, they must be obtained from two
+/// different calls to [FragmentProgram.fragmentShader].
+///
+/// Consider a fragment shader with uniforms like the following:
+///
+/// ```glsl
+/// layout (location = 0) uniform float a;
+/// layout (location = 1) uniform vec2 b;
+/// layout (location = 2) uniform vec3 c;
+/// layout (location = 3) uniform mat2x2 d;
+/// layout (location = 4) uniform sampler2d s;
+/// ```
+///
+/// If this shader is loaded into a [FragmentProgram] object `program`, a
+/// [FragmentShader] object can be constructed as follows:
+/// ```dart
+/// ImageShader sampler;
+/// final FragmentShader shader = program.fragmentShader()
+/// ..setFloat(0, 0.0)
+/// ..setFloat(1, 1.0)
+/// ..setFloat(2, 2.0)
+/// ..setFloat(3, 3.0)
+/// ..setFloat(4, 4.0)
+/// ..setFloat(5, 5.0)
+/// ..setFloat(6, 6.0)
+/// ..setFloat(7, 7.0)
+/// ..setFloat(8, 8.8)
+/// ..setFloat(9, 9.0)
+/// ..setSampler(0, sampler);
+///
+/// Paint paint = Paint()..shader = shader;
+/// ```
+///
+/// The uniforms will then be set as follows:
+///
+/// a: 0.0
+/// b: [1.0, 2.0]
+/// c: [3.0, 4.0, 5.0]
+/// d: [6.0, 7.0, 8.0, 9.0] // 2x2 matrix in column-major order
+///
+/// On subsequent frames, if only the uniform `a` is required to vary, then
+/// only `shader.setFloat(0, newValue)` needs to be called. The other uniforms
+/// will retain their values. The sampler uniforms will also persist in the same
+/// way.
+class FragmentShader extends Shader {
+  FragmentShader._(FragmentProgram program) : super._() {
+    _floats = _constructor(
+      program,
+      program._uniformFloatCount,
+      program._samplerCount,
+    );
+  }
+
+  static final Float32List _kEmptyFloat32List = Float32List(0);
+
+  late Float32List _floats;
+
+  /// Sets the float uniform at [index] to [value].
+  void setFloat(int index, double value) {
+    assert(!debugDisposed, 'Tried to accesss uniforms on a disposed Shader: $this');
+    _floats[index] = value;
+  }
+
+  /// Sets the sampler uniform at [index] to [sampler].
+  ///
+  /// All the sampler uniforms that a shader expects must be provided or the
+  /// results will be undefined.
+  void setSampler(int index, ImageShader sampler) {
+    assert(!debugDisposed, 'Tried to access uniforms on a disposed Shader: $this');
+    _setSampler(index, sampler);
+  }
+
+  /// Releases the native resources held by the [FragmentShader].
+  ///
+  /// After this method is called, calling methods on the shader, or attaching
+  /// it to a [Paint] object will fail with an exception. Calling [dispose]
+  /// twice will also result in an exception being thrown.
+  @override
+  void dispose() {
+    super.dispose();
```

Comment:
nit: call super.dispose last, since it might rip something out from under you

Reply:
Oh, hm, probably should call _dispose last since it deletes the native peer though...

Reply:
Right.